feat: Teach InfluxQL how to process `FILL(null|previous|<value>)`

stuartcarnie opened the pull request:
Helps #6916 

# Summary

Teaches IOx how to process the available gap-filling operations, using the features described by #6594:

* `FILL(null)`
* `FILL(previous)`
* `FILL(<value>)`

The `GapFill` operator is generated explicitly by the `InfluxQLToLogicalPlan` data type, rather that let the optimizer rewrite the plan, as it keeps all the concerns of InfluxQL within the InfluxQL crate. As a result of this decision, a number of the APIs for gap filling were made public per e6999a699d55049af0d7b2d8468ba791576ce43c. Also note that a small change was made to the `TimeRangeVisitor` type:

https://github.com/influxdata/influxdb_iox/blob/9afa91ec5704291212ee5c201f9d06c467993fb2/iox_query/src/logical_optimizer/handle_gapfill/range_predicate.rs#L31

to continue visiting the tree if it did not encounter one of the expected operators, allowing InfluxQL to provide a plan such as an aggregate, and it will continue:

https://github.com/influxdata/influxdb_iox/blob/9afa91ec5704291212ee5c201f9d06c467993fb2/iox_query/src/logical_optimizer/handle_gapfill/range_predicate.rs#L62

# Background

The planner was modified to look for the combination of `GROUP BY TIME` and one of the gap-filling strategies:

https://github.com/influxdata/influxdb_iox/blob/9afa91ec5704291212ee5c201f9d06c467993fb2/iox_query_influxql/src/plan/planner.rs#L481-L488

and if so, builds a `GapFill` operator over the aggregate operator:

https://github.com/influxdata/influxdb_iox/blob/9afa91ec5704291212ee5c201f9d06c467993fb2/iox_query_influxql/src/plan/planner.rs#L1039-L1046

> **Note**
>
> This implementation is derived from the following to suit InfluxQL behaviour:
>
> https://github.com/influxdata/influxdb_iox/blob/9afa91ec5704291212ee5c201f9d06c467993fb2/iox_query/src/logical_optimizer/handle_gapfill.rs#L136

## Bugs

Whilst increasing test coverage the following bugs were identified:

- #7361

wolffcm commented on influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:
```diff
@@ -668,10 +707,281 @@ Error while planning query: Error during planning: invalid number of arguments f
+| cpu              | 2022-10-31T02:00:30Z | 3     |
+| cpu              | 2022-10-31T02:01:00Z | 3     |
+| cpu              | 2022-10-31T02:01:30Z | 3     |
++------------------+----------------------+-------+
```
So because the fill value is a float, but the column type of `COUNT` is integer, it just truncates I guess? Seems reasonable.

stuartcarnie replied:
👍🏻 this is the same behaviour as InfluxQL

wolffcm reviewed (commented):
This looks great and it's really cool to see all this coming together.

My one concern is with your change to `range_predicate.rs`. I don't think it's quite right. I'd like to better understand what problem you were trying to solve there. I'm certain we can find a way to make it do what you need.

stuartcarnie commented on iox_query_influxql/src/plan/planner.rs:
```diff
@@ -1762,105 +1908,191 @@ mod test {
+                assert_snapshot!(plan("SELECT COUNT(f64_field) FROM data GROUP BY TIME(10s) FILL(previous)"), @r###"
+                Sort: time ASC NULLS LAST [iox::measurement:Dictionary(Int32, Utf8), time:Timestamp(Nanosecond, None);N, count:Int64;N]
+                  Projection: Dictionary(Int32, Utf8("data")) AS iox::measurement, time, COUNT(data.f64_field) AS count [iox::measurement:Dictionary(Int32, Utf8), time:Timestamp(Nanosecond, None);N, count:Int64;N]
+                    GapFill: groupBy=[[time]], aggr=[[COUNT(data.f64_field)]], time_column=time, stride=IntervalDayTime("10000"), range=Unbounded..Excluded(now()) [time:Timestamp(Nanosecond, None);N, COUNT(data.f64_field):Int64;N]
```
Yes, and I've even got a good workflow in CLion now that we have fully interactive terminal support!

stuartcarnie replied:
Updated the snapshot via 96569831e6649ab695da6062ac98769342911ece

stuartcarnie commented on iox_query/src/logical_optimizer/handle_gapfill/range_predicate.rs:
```diff
@@ -59,7 +59,7 @@ impl TreeNodeVisitor for TimeRangeVisitor {
             | LogicalPlan::Distinct(_) => Ok(VisitRecursion::Continue),
             // At some point we may wish to handle joins here too.
-            _ => Ok(VisitRecursion::Stop),
+            _ => Ok(VisitRecursion::Continue),
```
👍🏻 This is why I wanted to call this change out, as I suspected there was a good reason you specified `Stop` here.

I should have looked more closely at your usage at the call site, as I have done the same thing and it is fine. My issue was that I passed in the Aggregate plan rather than the input to the plan. I suppose the alternative is we could allow the visitor to traverse `Aggregate` operators?

stuartcarnie replied:
I have reverted that change via 50c6866d and made the adjustment in my code, and all the tests still pass 🥳

wolffcm replied:
> I suppose the alternative is we could allow the visitor to traverse Aggregate operators?

In this case the time column is passed into `DATE_BIN` so it would be tricky to track the column through `Aggregate`.

The GapFill operator takes care to apply `DATE_BIN` to the bounds of the range to ensure that we get the first and last timestamps correct:
https://github.com/influxdata/influxdb_iox/blob/main/iox_query/src/exec/gapfill/params.rs#L74-L81

So logically to me it makes the most sense to find the range as the input to `Aggregate`.

stuartcarnie commented on iox_query_influxql/src/plan/planner.rs:
```diff
@@ -475,6 +476,48 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
+            let args = match select_exprs[time_column_index].clone().unalias() {
+                Expr::ScalarFunction {
+                    fun: BuiltinScalarFunction::DateBin,
+                    args,
```
It doesn't, as I don't use it in this iteration. I was in the first experiment and letting the optimiser do the rewriting.

Turns out this approach is easier, as all the gap-filling logic is in one function (`select_aggregate`) where as previously, I was going to have to modify other areas to add `LOCF`, etc to the projection.

stuartcarnie commented on iox_query_influxql/src/plan/planner.rs:
```diff
@@ -983,6 +1036,94 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
+        )));
+    };
+
+    let aggr = Aggregate::try_from_plan(&input)?;
```
No, this is just unwrapping `input`, which is expected to be an `Aggregate` plan created by the caller, `select_aggregate`.

stuartcarnie commented:
Thanks for the review @wolffcm – I've reverted the change to the `find_time_range` function and adjusted my code accordingly and all the tests are still good.

wolffcm reviewed (approved):
Looks good!
